digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/client/vvh/KgvOperation.scala

```diff
@@ -0,0 +1,493 @@
+    val request = new HttpGet(url)
+    addHeaders(request)
+    
+    val client =  HttpClients.custom()
+                  .setDefaultRequestConfig( RequestConfig.custom()
+                                  .setCookieSpec(CookieSpecs.STANDARD)
+                                  .build())
+                  .build()
+    
+    var response: CloseableHttpResponse = null
+    LogUtils.time(logger, s"fetch roadLink features",url = Some(url)) {
+      try {
+        response = client.execute(request)
+        val statusCode = response.getStatusLine.getStatusCode
+        if (statusCode == HttpStatus.SC_OK) {
+          val feature = parse(StreamInput(response.getEntity.getContent)).values.asInstanceOf[Map[String, Any]]
+          val resort = feature("type").toString match {
+            case "Feature" => 
+              if (roadLinkStatusFilter(feature)){
+                Some(FeatureCollection(
+                  `type` = "FeatureCollection",
+                  features = List(convertToFeature(feature)),
+                  crs = Try(Some(feature("crs").asInstanceOf[Map[String, Any]])).getOrElse(None)))
+              } else None
+            case "FeatureCollection" =>
+              val links = feature("links").asInstanceOf[List[Map[String, Any]]].map(link=>
+                Link(link("title").asInstanceOf[String], link("type").asInstanceOf[String],
+                  link("rel").asInstanceOf[String], link("href").asInstanceOf[String])
+              )
+              
+              val nextLink = Try(links.find(_.title=="Next page").get.href).getOrElse("")
+              val previousLink = Try(links.find(_.title=="Previous page").get.href).getOrElse("")
+              val features = feature("features").asInstanceOf[List[Map[String, Any]]] .filter(roadLinkStatusFilter)
+                                                                                      .map(convertToFeature)
+              Some(FeatureCollection(
+                `type` = "FeatureCollection",
+                features = features,
+                crs = Try(Some(feature("crs").asInstanceOf[Map[String, Any]])).getOrElse(None),
+                numberReturned = feature("numberReturned").asInstanceOf[BigInt].toInt,
+                nextLink,previousLink
+              ))
+            case _ => None
+          }
+          Left(resort)
+        } else {
+          Right(LinkOperationError(response.getStatusLine.getReasonPhrase, response.getStatusLine.getStatusCode.toString))
+        }
+      }
+      catch {
+        case e: Exception => Right(LinkOperationError(e.toString, ""))
+      }
+      finally {
+        if (response != null) {
+          response.close()
+        }
+      }
+    }
+  }
+  
+  def paginationRequest(base:String,limit:Int,startIndex:Int = 0,firstRequest:Boolean = true ): (String,Int) = {
+    if (firstRequest) (s"${base}&limit=${limit}&startIndex=${startIndex}",limit)
+    else (s"${base}&limit=${limit}&startIndex=${startIndex}",startIndex+limit)
```

Comment:
Miten nämä startIndex käytännössä menee? Jos ensiksi haetaan 0-4999 linkkiä ja seuraavaksi haettaisiin 4999-9998 niin olisiko ensimmäisen haun viimeinen linkki sama kuin toisen haun ensimmäinen eli ns. 4999s linkki? Eli pitäisikö seuraavan position olla kuitenkin startIndex+limit+1 ?

Reply:
Uudelleen mietittynä taitaa toimia oikein näin kun tarkastelu aloitetaan nollasta.

Reply:
hyym eikös toi ole sama kuin tossa 394 limit *2, eikös +1 tarkoita että hypätään yksi linkki